### ISSUE TITLE
Moderation of class and student lists

### ISSUE THREAD

**Issue by pricem** [learning-unlimited/ESP-Website#1229]
It is desired to require admin approval of e-mails sent to forwarding lists such as M100-students and M100-teachers.  While this capability is allowed by Mailman, it may need to be added to the built-in existing e-mail forwarding system.  (Requested by Stanford in spring 2014.)

See also #119 and #298.

**Comment by willgearty** [comment]
Stanford is having some problems with spam emails to class lists, in addition to some inappropriate material being shared from teachers to students.
Any chance this or #1680 is getting any attention?
@pricem @benjaminjkraft ?

**Comment by milescalabresi** [comment]
Other chapters have also requested that /admin/dbmail/textofemail/ get a copy of these messages.

**Comment by benjaminjkraft** [comment]
@willgearty I think as a workaround they do already cc the server archive address.  That's currently a shared LU archive address, but if you want you could create a gmail and ask websupport to redirect it there.  (It might be a good idea to also give web-team access to that archive for debugging, although it's not strictly necessary.)  It gets a lot of email, so you won't want that to directly go to a bunch of people, but you could set up a filter to forward just the class mailing list traffic to another list or person.

**Comment by willgearty** [comment]
@benjaminjkraft Yeah, our gmail account already gets a copy of all of the emails. But I do think there is merit to having them archived on the website as well, as @milescalabresi suggested.
However, our real problem is being able to moderate the emails BEFORE they are sent to the students/teachers/etc.
Some way to approve or reject emails before the server then sends them to the recipients.
As I said in my email to websupport, our chapter had a pretty severe incident this past Splash via these mailing lists and we had no control over what the teacher was able to send to his students.

**Comment by benjaminjkraft** [comment]
Ok, got it.

**Comment by willgearty** [comment]
Just discovered there is an `admin_hold` field on the `EmailList` model. It's never used in the codebase, but it seems like someone at least thought this feature would be useful. It would require some sort of UI for admins to moderate emails.

**Comment by abhinavjha0239** [comment]
Hi @willgearty , I would like to work on this. Can you please assign this to me.

**Comment by abhinavjha0239** [comment]
After building the moderation system, I've been stress-testing the design and wanted to flag some edge cases and get input on a few design questions.

## Edge cases I've identified

**Teacher experience (most critical):**
1. **Teacher has no idea their email is held.** Teacher sends email, assumes it was delivered. Students never got it. We currently send zero notification to the teacher.
2. **Reply chains break on list addresses.** Student replies to an approved class-list email. Teacher replies back via the list. Teacher's reply gets held again. Student thinks teacher ghosted them.
3. **Teacher resends thinking it failed.** No delivery confirmation → teacher sends the same email 3 times → admin sees 3 duplicates in queue. (Addressed by Proposal 1 — once teacher gets "held for review" notification, they won't resend.)

**Timing problems:**
4. **Admin is late / unavailable.** Teacher sends "Class moved to Room 205, starts in 30 min" at 11:30 AM. Admin checks at 3 PM. Students went to the wrong room.
5. **Off-hours emails.** Teacher sends "Tomorrow's class is cancelled" Friday night. Admin sees it Monday morning — useless by then.
6. **No admin on duty.** All admins away. Emails pile up indefinitely with no expiry, no escalation.

**Admin experience:**
7. **Accidental reject — no undo.** Admin clicks "Reject" by mistake. The email data is preserved in the database but there's no path to approve a rejected email — the teacher has to resend.
8. **Stale content on late approval.** Email held for 3 days says "submit homework by tomorrow." Admin approves — now "tomorrow" means the wrong date.

## Proposals

### 1. Teacher notification + "deliver by" deadline

When an email is held, send an auto-reply to the teacher:

> *Your email to [class list] is being held for admin review.*
> *If you need this delivered by a specific time, you can set a deadline here: [link]*

The link goes to a small authenticated page where the teacher can set a "deliver by" datetime (using a token-based URL so the teacher doesn't need to be logged in — similar to how unsubscribe links already work in the codebase). This deadline gets stored on the `HeldEmail` and shown to the admin in the moderation queue alongside the email, so the admin can prioritize time-sensitive messages.

This also connects naturally with the teacher comm panel (3833) — if teachers compose via the web interface instead of their email client, we can show the moderation status inline and include a "deliver by" picker directly in the compose form, no auto-reply needed.

| | Mailgate path (email client) | Teacher comm panel (3833) |
|---|---|---|
| Moderation notice | Auto-reply email | Banner in compose UI |
| Deliver-by deadline | Link in auto-reply → web page | Datetime picker in form |
| Scheduled send + moderation | N/A (no scheduling on mailgate) | Teacher schedules for March 10 12 PM → email enters moderation queue immediately → admin reviews before scheduled time → sends at 12 PM |

### 2. Auto-approve on deadline expiry

For edge cases #4–6: if the teacher's "deliver by" deadline passes without admin action, auto-approve and send. This prevents moderation from becoming a communication black hole when admins are unavailable.

Teachers have been emailing class lists directly for 20 years with no review — auto-approve after a deadline is still strictly better than the current baseline (zero moderation). Chapters that want stricter control can disable auto-approve or set a longer fallback timeout.

Configuration (Tag-based, per-program):
- `moderation_auto_approve_timeout`: fallback timeout when no teacher deadline is set (default: 24h, 0 to disable)
- Teacher-set deadlines override the fallback when shorter

When auto-approve fires, notify the admin: *"Email from [teacher] to [class list] was auto-approved after [X hours] — no moderator acted before the deadline."*

Implementation: a small periodic check (either added to `dbmail_cron.py` or a separate lightweight cron) that queries `HeldEmail.objects.filter(status='pending', deliver_by__lte=now)` and auto-approves.

### 3. Rejection notification to teacher

For edge case #7: when an admin rejects an email, notify the teacher with the rejection reason — *"Your email to [class list] was not approved. Reason: [reason]. You can revise and resend."* Teacher knows what happened and can fix and resend instead of the email silently disappearing.

### 4. Escalation notifications

If emails are pending for more than a configurable threshold (e.g. 2 hours), notify admins. If still unmoderated after a longer threshold (e.g. 6 hours), escalate to directors.

## Note on scheduling interaction

One thing I noticed — the moderation system (mailgate → `HeldEmail._send()` → direct sendmail) and the comm panel (`MessageRequest` → cron → `TextOfEmail` → `send_mail()`) are completely separate delivery pipelines. Moderated emails bypass the cron entirely and don't create `TextOfEmail` records, which also means they're invisible to email logging (#298).

This isn't blocking for the current PR, but worth keeping in mind — at some point we may want approved moderated emails to optionally route through the cron for scheduling and logging.

---

Thoughts? Happy to implement the teacher notification + deadline in this PR if the direction makes sense.

---

The core moderation system (hold, review, approve/reject) is already implemented and open for review in #4917 — the proposals above are forward-looking additions on top of that foundation.
